Adds support for go worker in breeze start

aritra24 opened the pull request:
Adds a --worker-type command to breeze start-airflow command
which launches workers of worker-type (only golang so far) in a separate tab in breeze.

closes: #51947

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

aritra24 committed:
Draft: Added support for go worker

closes: #51947

aritra24 committed:
Merge branch 'main' into add_go_worker_to_breeze

aritra24 committed:
Merge branch 'main' into add_go_worker_to_breeze

aritra24 committed:
Merge branch 'main' into add_go_worker_to_breeze

aritra24 committed:
Updated command to run go worker

aritra24 committed:
Fixed up the tmux command

aritra24 committed:
Merge branch 'main' into add_go_worker_to_breeze

gopidesupavan commented on scripts/in_container/bin/run_tmux:
```diff
@@ -123,6 +123,17 @@ if [[ ${STANDALONE_DAG_PROCESSOR} == "true" ]]; then
+    tmux split-window -h
+    tmux set-option -p @airflow_component "Go worker"
+    tmux send-keys "cd go-sdk" C-m
+    tmux send-keys "go run ./example/main.go run \
```
do we need this run?

amoghrajesh replied:
Yeah we need it. The second run it the "run" subcommand for go workers

amoghrajesh commented on dev/breeze/src/airflow_breeze/commands/common_options.py:
```diff
@@ -525,3 +525,10 @@ def _normalize_platform(ctx: click.core.Context, param: click.core.Option, value
     type=BetterChoice(SINGLE_PLATFORMS),
 )
+
+option_go_worker = click.option(
```
Set default to false?

amoghrajesh commented on dev/breeze/src/airflow_breeze/commands/developer_commands.py:
```diff
@@ -625,6 +631,7 @@ def start_airflow(
         force_build=force_build,
         forward_credentials=forward_credentials,
         github_repository=github_repository,
+        go_worker=go_worker,
```
We should probably add one check though, we should validate if the executor is `CeleryExecutor` or fail early

amoghrajesh reviewed (commented):
Mostly looks nice @aritra24, some comments.

potiuk commented on scripts/in_container/bin/run_tmux:
```diff
@@ -123,6 +123,17 @@ if [[ ${STANDALONE_DAG_PROCESSOR} == "true" ]]; then
     tmux send-keys 'airflow dag-processor' C-m
 fi
 
+if [[ ${STANDALONE_DAG_PROCESSOR} == "true" ]]; then
```
```suggestion
if [[ ${WORKER} == "go" ]]; then
```

aritra24 committed:
Merge branch 'main' into add_go_worker_to_breeze

aritra24 committed:
Adds support for go worker on breeze

Adds new params --worker-types which currently
only takes in a single type of go and then starts
the go worker.

aritra24 committed:
Adds executor check

Early exit if edge executor isn't enabled

aritra24 commented on go-sdk/edge/worker.go:
```diff
@@ -121,10 +121,10 @@ func NewWorker(
 	airflowVer.FromWorkerStateBodySysinfo0(edgeapi.WorkerStateBodySysinfo0(
-		configOrDefault("edge.airflow_version", "3.1.0"),
+		configOrDefault("edge.airflow_version", "3.2.0"),
 	))
 	edgeVer.FromWorkerStateBodySysinfo0(edgeapi.WorkerStateBodySysinfo0(
-		configOrDefault("edge.provider_version", "1.3.1"),
+		configOrDefault("edge.provider_version", "3.2.0"),
```
I changed this to bring it into sync and make it work. Wasn't sure what the specific criteria for updating it is, so can definitely revert this one.

aritra24 commented on go-sdk/pkg/edgeapi/client.go:
```diff
@@ -39,6 +39,7 @@ func WithEdgeAPIJWTKey(key []byte) ClientOption {
 			now := time.Now().UTC().Unix()
 			t := jwt.NewWithClaims(jwt.SigningMethodHS512, jwt.MapClaims{
 				"method": endpointPath,
+				"iss":    "airflow",
 				"aud":    "api",
```
This one needs fixing, needs to come from config. I added it temporarily since without the issuer the validating was failing for me. @ashb maybe you can comment?

aritra24 commented on scripts/in_container/bin/generate_mprocs_config.py:
```diff
@@ -62,6 +62,10 @@ def generate_mprocs_config() -> str:
 
+    API_SECRET_KEY = ""
+    if get_env("GO_WORKER") != "":
+        API_SECRET_KEY = "4cT8uMo04i5fOFoSICIMnw=="
```
Using a hardcoded value since it needs to be shared between the go worker and the api server? Not fully sure how the edge worker can get the value otherwise? Open to ideas

jscheffl replied:
Please no hard coded secrets. Even not temporarily.

In Edge also the ENV is used and usually temporal secrets are generated by Breeze, so you could do the same like Edge worker uses for startup.

aritra24 replied:
Ack, I was confused on how the python worker was getting it too. Could you point to where the environment var is getting set/shared? I can maybe hook in setting the other environment vars in the same place.

jscheffl replied:
dev/breeze/src/airflow_breeze/params/shell_params.py:587

aritra24 replied:
Thanks, I'll check it out tomorrow and update the pr with all the changes!

aritra24 replied:
resolved this

uranusjr commented on dev/breeze/src/airflow_breeze/commands/common_options.py:
```diff
@@ -602,6 +603,13 @@ def _normalize_platform(ctx: click.core.Context, param: click.core.Option, value
 
+option_worker_types = click.option(
+    "--worker-types",
+    help="Start a specific worker",
```
This should be `--worker-type` (or maybe just `--worker`) since each flag only takes one worker type (e.g. `breeze --worker=go --worker=java`).

aritra24 replied:
Ack, I think worker-type is better since there's already a workers which can cause confusion. 
```
"--workers",
    help="Number of parallel workers for e2e tests",
```

aritra24 replied:
done

aritra24 commented on scripts/in_container/bin/generate_mprocs_config.py:
```diff
@@ -119,6 +121,20 @@ def generate_mprocs_config() -> str:
 
+    if get_env("GO_WORKER") != "":
+        env = {}
+        env["AIRFLOW__EDGE__API_URL"] = "http://localhost:8080"
+        env["AIRFLOW__BUNDLES__FOLDER"] = "./bin"
```
Hardcoded these two for now, can take them from variables/configs as a follow up

uranusjr commented on dev/breeze/src/airflow_breeze/global_constants.py:
```diff
@@ -170,11 +170,14 @@
 START_AIRFLOW_ALLOWED_EXECUTORS = [LOCAL_EXECUTOR, CELERY_EXECUTOR, EDGE_EXECUTOR]
 START_AIRFLOW_DEFAULT_ALLOWED_EXECUTOR = START_AIRFLOW_ALLOWED_EXECUTORS[0]
 ALLOWED_CELERY_EXECUTORS = [CELERY_EXECUTOR, CELERY_K8S_EXECUTOR]
+ALLOWED_WORKER_TYPES = [GOLANG_WORKER]
```
This can probably also support running the Python worker (`airflow edge worker`)?

aritra24 replied:
Yup, can do that. I think today it launches the python worker by default if we use edge executor. I can see if I can break that out into using the worker-type flag instead. 🤔

jscheffl replied:
Edge worker is today started per default if `EdgeExecutor` is selected.

aritra24 replied:
@jscheffl what do you think of splitting out the edge worker startup out into a separate worker type flag? I can try breaking that out as part of a follow up once this goes in?

jscheffl replied:
Why should we? I think if you want to test EdgeExecutor it would be always required to have one worker. I see low benefit of adding this to an extra parameter because there would be almost no use case starting without.

So I'd say: No... until there is a benefit in this.

In my view a Go worker is something additional but I see no real use case where you only start a Go worker and no other.

jscheffl replied:
I'd therefore also rather propose to call the parameter "Additional Worker". Because assuming you always have a Python based worker and then potentially you want to start a Go and a Java worker additionally?

aritra24 replied:
I don't have strong opinions on it so I'm okay with leaving it as is, but if we're only testing a go based dag would we need the python worker running?

jscheffl replied:
Then _all_ tasks need to be written in Go and you need to restart if you want to test something else. If this is a valid use case... I am not sure and would assume it is always rather hybrid.

uranusjr replied:
I feel running all tasks in Go is a very valid use case _in Breeze_ specifically.

aritra24 replied:
I do think it's a valid use case as well, especially during dev/testing of the SDK itself where'd I'm mostly be working with go based tasks only.

jscheffl replied:
So then your aim is really to fully replace all "normal" worker types with a Go worker?

Then same yields for Celery worker. LocalExecutor excempted maybe. So problably you talk about "Python Worker" which then refers to either Celery or Edge worker.

jscheffl replied:
Any Python worker should be the default.

uranusjr replied:
Of course, a Python worker should be the default in any cases. I think the use case is only specifically to test-run functionalities in the Go SDK. Currently, the Go SDK requires you to use EdgeExecutor, and have a Go worker process running. The Python worker is not needed to execute Go tasks.

It’s a very specific use case that only applies to dev (with Breeze), and honestly not even necessary (I can just leave the Python worker running idly). It would be good enough if there’s a way to automatically start the Go worker when EdgeExecutor is used.

What I would want ideally

```bash
--executor EdgeExecutor  # Start a Python worker automatically.
--executor EdgeExecutor --worker-type Python  # Start a Python worker (explicit).
--executor EdgeExecutor --worker-type Go  # Start (only) a Go worker.
--executor EdgeExecutor --worker-type Python --worker-type Go  # Start two workers.
```

I don’t really mind CeleryExecutor for now (I believe the Go worker does not work with it currently), but we could do the same and add support later (I believe being able to run foreign language SDK in other executors is a planned feature anyway).

aritra24 replied:
deferring this for now, will look into adding this as a follow up in some time, since there's already a bunch of stuff in this pr.

jscheffl commented on scripts/docker/entrypoint_ci.sh:
```diff
@@ -136,6 +136,13 @@ function environment_initialization() {
         export AIRFLOW__SCHEDULER__STANDALONE_DAG_PROCESSOR=True
     fi
 
+    if [[ ${GO_WORKER=} == "true" ]]; then
+        echo
+        echo "${COLOR_BLUE}Starting go worker${COLOR_RESET}"
+        echo
+        export AIRFLOW__SCHEDULER__GO_WORKER=True
+    fi
```
Note that Celery and Edge workers are started in different places. In my view would make sense to move the starting point to the same... whereas I am not sure whether I got the right context. At least I assume it is better to start in tmux/mproc panels?

aritra24 replied:
This one's only setting the variable to trigger the worker, it's started in tmux/mprocs. Unless I've misunderstood your point? Though I don't think I use this variable anymore it's from an older commit so can probably be removed 🤔 let me check.

jscheffl replied:
Ah, you are right, only an export, not a start.

aritra24 committed:
Updates mprocs config to remove hardcoded secrets

jscheffl reviewed (approved):
Looks good for me

aritra24 committed:
Merge branch 'main' into add_go_worker_to_breeze

aritra24 committed:
Uses the new console_print instead of older get_console.print

aritra24 committed:
Merge branch 'main' into add_go_worker_to_breeze

aritra24 committed:
Merge branch 'main' into add_go_worker_to_breeze

aritra24 committed:
Merge branch 'main' into add_go_worker_to_breeze

aritra24 committed:
Merge branch 'main' into add_go_worker_to_breeze

aritra24 committed:
Updates docs

aritra24 committed:
Merge branch 'main' into add_go_worker_to_breeze

aritra24 committed:
Merge branch 'main' into add_go_worker_to_breeze

aritra24 committed:
Merge branch 'main' into add_go_worker_to_breeze

aritra24 committed:
Merge branch 'main' into add_go_worker_to_breeze

potiuk reviewed (approved):
Nice!
